src/coap_oscore.c

```diff
@@ -0,0 +1,2173 @@
+      !coap_insert_option(pdu,
+                          COAP_OPTION_URI_PORT,
+                          coap_encode_var_safe(option_value_buffer,
+                                               sizeof(option_value_buffer),
+                                               uri.port & 0xffff),
+                          option_value_buffer))
+    goto error;
+  if (uri.path.length) {
+    uint8_t *buf;
+    uint8_t *kbuf;
+    size_t buflen = uri.path.length + 1;
+    int res;
+
+    kbuf = buf = coap_malloc(uri.path.length + 1);
+    if (buf) {
+      res = coap_split_path(uri.path.s, uri.path.length, buf, &buflen);
+      while (res--) {
+        if (!coap_insert_option(pdu,
+                                COAP_OPTION_URI_PATH,
+                                coap_opt_length(buf),
+                                coap_opt_value(buf))) {
+          coap_free(buf);
+          goto error;
+        }
+        buf += coap_opt_size(buf);
+      }
+    }
+    coap_free(kbuf);
+  }
+  if (uri.query.length) {
+    uint8_t *buf;
+    size_t buflen = uri.query.length + 1;
+    int res;
+
+    buf = coap_malloc(uri.query.length + 1);
+    if (buf) {
+      res = coap_split_query(uri.query.s, uri.query.length, buf, &buflen);
+      while (res--) {
+        if (!coap_insert_option(pdu,
+                                COAP_OPTION_URI_QUERY,
+                                coap_opt_length(buf),
+                                coap_opt_value(buf))) {
+          coap_free(buf);
+          goto error;
+        }
+
+        buf += coap_opt_size(buf);
+      }
+      coap_free(buf);
+    }
+  }
+  if (!coap_insert_option(pdu,
+                          COAP_OPTION_PROXY_SCHEME,
+                          strlen(coap_uri_scheme[uri.scheme]),
+                          (const uint8_t *)coap_uri_scheme[uri.scheme]))
+    goto error;
+  coap_free(keep_proxy_uri);
+  return 1;
+
+error:
+  coap_free(keep_proxy_uri);
+  return 0;
+}
+
+static void
+dump_cose(cose_encrypt0_t *cose, const char *message) {
+  if (coap_get_log_level() >= COAP_LOG_OSCORE) {
+    char buffer[30];
+
+    coap_log_oscore("%s Cose information\n", message);
+    oscore_log_char_value(COAP_LOG_OSCORE, "alg",
+                          cose_get_alg_name(cose->alg, buffer, sizeof(buffer)));
+    oscore_log_hex_value(COAP_LOG_OSCORE, "key", &cose->key);
+    oscore_log_hex_value(COAP_LOG_OSCORE, "partial_iv", &cose->partial_iv);
+    oscore_log_hex_value(COAP_LOG_OSCORE, "key_id", &cose->key_id);
+    oscore_log_hex_value(COAP_LOG_OSCORE, "kid_context", &cose->kid_context);
+    oscore_log_hex_value(COAP_LOG_OSCORE,
+                         "oscore_option",
+                         &cose->oscore_option);
+    oscore_log_hex_value(COAP_LOG_OSCORE, "nonce", &cose->nonce);
+    oscore_log_hex_value(COAP_LOG_OSCORE, "external_aad", &cose->external_aad);
+    oscore_log_hex_value(COAP_LOG_OSCORE, "aad", &cose->aad);
+  }
+}
+
+#define MAX_IV_LEN 10 /* maximum length of iv buffer */
```

Comment:
The COSE definitions suggest that the (AEAD) IV can be up to 13 bytes. Is this a different type of IV?

Reply:
Actually, it is used for "Partial IV" which is used for the sender sequence number (8 bytes).  Re-ordered code variable location and size to make this clearer.